tests/codegen.rs

```diff
@@ -1,118 +1,71 @@
-            continue;
-        }
-
-        certs.push(current);
-        current = Vec::with_capacity(16);
-    }
-
-    certs.push(current);
-
-    // Parse each certificate and check fingerprints
-
-    let mut hashed = HashMap::<Vec<u8>, (Vec<&str>, Vec<u8>)>::default();
-    for cert in certs {
-        let start = cert
-            .iter()
-            .position(|&ln| ln == "-----BEGIN CERTIFICATE-----")
-            .unwrap();
-
-        // Parse PEM to get the DER.
-        let pem = cert[start..].join("\n");
-        let mut reader = Cursor::new(pem.as_bytes());
-        let der = match rustls_pemfile::read_one(&mut reader).unwrap().unwrap() {
-            rustls_pemfile::Item::X509Certificate(der) => der,
-            _ => unreachable!(),
-        };
-
-        // Check if our hash matches the one in the file.
-        let their_fingerprint = cert
-            .iter()
-            .find_map(|&ln| ln.strip_prefix("# SHA256 Fingerprint: "))
-            .unwrap();
-        let hash = digest::digest(&digest::SHA256, &der);
-        assert!(
-            !hashed.contains_key(hash.as_ref()),
-            "duplicate hash: {:#?}",
-            &cert
-        );
-
-        let our_fingerprint = hash
-            .as_ref()
-            .iter()
-            .map(|b| format!("{:02x}", b))
-            .collect::<String>();
-        assert_eq!(
-            their_fingerprint.replace(':', ""),
-            our_fingerprint,
-            "{:#?}",
-            &cert
-        );
-
-        hashed.insert(hash.as_ref().to_vec(), (cert, der));
-    }
-
-    // For the given certificate subject name, store a name constraints encoding
-    // which will be applied to that certificate. This data is sourced from
-    // https://hg.mozilla.org/projects/nss/file/tip/lib/certdb/genname.c such that
-    // webpki-roots implements the same policy in this respect as the Mozilla root program.
-    let mut imposed_constraints = HashMap::<Vec<u8>, Vec<u8>>::default();
-    imposed_constraints.insert(
-        concat(TUBITAK1_SUBJECT_DN),
-        TUBITAK1_NAME_CONSTRAINTS.to_vec(),
-    );
-
-    // Generate the trust anchors, sorted by fingerprint
-
-    let mut hashes = hashed.into_iter().collect::<Vec<_>>();
-    hashes.sort_by(|a, b| a.0.cmp(&b.0));
-
-    let (mut subject, mut spki, mut name_constraints) =
-        (String::new(), String::new(), String::new());
     let mut code = String::with_capacity(256 * 1_024);
     code.push_str(HEADER);
     code.push_str("pub const TLS_SERVER_ROOTS: &[TrustAnchor] = &[\n");
-    for (_, (lines, der)) in hashes {
-        let ta = webpki::TrustAnchor::try_from_cert_der(&der).unwrap();
+    let (mut subject, mut spki, mut name_constraints) =
+        (String::new(), String::new(), String::new());
+
+    for root in trusted_tls_roots {
+        // Verify the DER FP matches the metadata FP.
+        let der = root.der();
+        let calcualted_fp = digest::digest(&digest::SHA256, &der);
```

Comment:
`calcualted_fp` is probably not what you were going for...

Reply:
maebe knot :laughing: 

Fixed, thanks.